About page, configurable cli download links

fabianofranz opened the pull request:
Adds an "About" page to the web console with server version information and configurable download links for the CLI tools (if provided).

I wanted to make the download URL's flexible enough to allow you to either specify multiple links pointing directly to the client packages (for each platform, for example), or you could point to an external page that links the actual packages (the GitHub releases page, for example). So the configuration takes both the URL and title. Examples of valid configurations would be:

```
cli: {
  downloadURL: {
    "Linux (32 bits)": "https://<cdn>/cli-v1.1.1-e1d9873-linux-32bit.tar.gz",
    "Linux (64 bits)": "https://<cdn>/cli-v1.1.1-e1d9873-linux-64bit.tar.gz",
    "Windows": "https://<cdn>/cli-v1.1.1-e1d9873-windows.zip",
    "Mac OS X": "https://<cdn>/cli-v1.1.1-e1d9873-mac.zip"
  }
}
```

or

```
cli: {
  downloadURL: {
    "Latest Release": "https://<cdn>/openshift-origin-client-tools/releases"
  }
}
```

About page

![openshift web console](https://cloud.githubusercontent.com/assets/158611/12519445/f7a36bd0-c125-11e5-894c-a983fc211cee.png)


Fixes https://github.com/openshift/origin/issues/2281
Fixes https://github.com/openshift/origin/issues/3493
Fixes https://github.com/openshift/origin/issues/4344
Fixes https://github.com/openshift/origin/issues/4885

liggitt commented on pkg/cmd/server/api/v1/types.go:
```diff
@@ -377,6 +377,9 @@ type AssetConfig struct {
 	// MetricsPublicURL is the public endpoint for metrics (optional)
 	MetricsPublicURL string `json:"metricsPublicURL"`
 
+	// CLIDownloadPublicURLs holds links to download client tools, where key is the platform name and value is the download URL (optional)
```
Hmm, this seems super specific. Won't there be other things that need customizing on that page as well, like product name and other links? Plumbing all of that through config seems like the hard way to do that

fabianofranz replied:
We may want to reduce the amount of info in that page, but an entry in the config file for CLI download links is a specific requirement from online. Links for docs already use the filters mechanism used everywhere on the web console.

jwforres replied:
@liggitt we've talked with @brenton and @danmcp about specifically letting administrators put in their own download links so they can host `oc` themselves, so this is more than just per product config, its really administrative config

liggitt replied:
we also don't expose doc links in config... I would expect doc links and cli download links to use the same mechanism

fabianofranz replied:
I think this is the only thing left in this PR, waiting for @jwforres confirmation about where we want this config.

fabianofranz commented:
@jwforres @spadgett

jwforres commented:
try the openshift icon in red, check with jluc

jwforres commented:
... hit enter too soon...

check with @luciddreamz for what the right red is for the openshift icon

jwforres commented:
the download links blend together a bit, maybe stacking them vertically, or having a gray pipe between them

spadgett commented:
@sspeiche FYI

sspeiche commented on assets/app/views/about.html:
```diff
@@ -0,0 +1,65 @@
+              </div>
+
+              <h2>About</h2>
+              <p><a href="https://openshift.com">OpenShift</a> is Red Hat's Platform-as-a-Service (PaaS) that allows developers to quickly develop, host, and scale applications in a cloud environment.</p>
```
I wonder, if this is the about page for Origin, then we may want to have the link go to openshift.org
Likewise, at some point we may want to specific links to the offering page for Enterprise and Dedicated?.

Though I think for an initial pass, it would be fine just defaulting to openshift.com as it has redirects already

fabianofranz replied:
Yeah, at least for now the idea was to point to openshift.com which links every other product and project.

sspeiche commented:
It all looks great. Only thing I'm not sure of is how we are directing users to get help, provide feedback or report problems.  Would this come in the about page or a separate menu item or ?

liggitt commented on pkg/cmd/server/origin/asset.go:
```diff
@@ -218,6 +220,20 @@ func (c *AssetConfig) addHandlers(mux *http.ServeMux) error {
+	if err != nil {
+		return err
+	}
+	versionPath := path.Join(publicURL.Path, "version.js")
```
does this have to be served in a separate file? don't really want lots of individual requests for config data

jwforres replied:
agree with @liggitt why wouldn't this be in config.js ?

fabianofranz replied:
All `config.js` entries are something you can change to customize some behavior, while version info is always generated by the master, not part of the set of configs an installation could change. Didn't seem to fit well. But I could buy having it in config to save a request, I think.

fabianofranz replied:
Fixed, version is now being server in `config.js`, in a separate object as it was before to keep what's really configurable as a separate structure.

fabianofranz commented:
> the download links blend together a bit, maybe stacking them vertically, or having a gray pipe between them

@jwforres yeah, my first version was vertical and for some reason I changed, moved back which looks better

fabianofranz commented:
> check with @luciddreamz for what the right red is for the openshift icon

@luciddreamz using `#c00` until we confirm what's the right color

fabianofranz commented:
> It all looks great. Only thing I'm not sure of is how we are directing users to get help, provide feedback or report problems. Would this come in the about page or a separate menu item or ?

@sspeiche I *think* the design review in the works by @sg00dwin includes that, it has a "?" drop-down that include links to docs, this About page, and probably other stuff. Anyone else could confirm?

luciddreamz commented:
@fabianofranz the OpenShift logo is #CD202C:
![screen shot 2016-01-22 at 1 21 03 pm](https://cloud.githubusercontent.com/assets/928421/12519260/c1b0cb86-c10b-11e5-95d7-55ff5ee61f65.png)

jwforres commented:
There would be a place to add other links there, we need to think about
"Report a problem" we had a lot of issues trying to do that in v2 because
different products and even different user plans had different ways to
report problems.

On Friday, January 22, 2016, Fabiano Franz <notifications@github.com> wrote:

> It all looks great. Only thing I'm not sure of is how we are directing
> users to get help, provide feedback or report problems. Would this come in
> the about page or a separate menu item or ?
>
> @sspeiche <https://github.com/sspeiche> I *think* the design review in
> the works by @sg00dwin <https://github.com/sg00dwin> includes that, it
> has a "?" drop-down that include links to docs, this About page, and
> probably other stuff. Anyone else could confirm?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6783#issuecomment-174000204>.
>

fabianofranz commented:
@luciddreamz thanks! updated

jwforres commented:
Ok result of a number of discussions is that we are going to do this as an extension JS.  So we need a JS API for providing custom download links, think about how you would do the same thing for help links (and if possible go ahead and implement it cause we need it soon).

OOTB in origin I'm guessing we wouldn't provide a CLI download link since the only place they exist today is in the release in github which includes ALL the binaries.

So if we dont have any CLI download links available, then maybe tweak the text in the CLI section?

fabianofranz commented:
@jwforres ok, thanks, I'll adjust this to use JS extensions. No CLI download links available were already a supported scenario, the About page adjusts to it.
@spadgett do we already use extension JS somewhere in the code? Could you point me to it?

spadgett commented:
> Ok result of a number of discussions is that we are going to do this as an extension JS.

It seems odd to ask admins to write JS code to set one URL :-/

spadgett commented:
> do we already use extension JS somewhere in the code? Could you point me to it?

Not yet, but here is the help topic:

https://docs.openshift.org/latest/install_config/web_console_customization.html

Let me know if you have trouble.

fabianofranz commented:
@spadgett thanks!

fabianofranz commented:
@jwforres @spadgett Ok, so please take a look. 

The proposal is a central service that handles [several constants](https://github.com/fabianofranz/origin/blob/484_download_cli_from_console/assets/app/scripts/constants.js) which can be overridden individually or entirely by JS extension(s). Example of an extension that overrides the CLI download links to point to individual packages:

```
window.CONSTANTS.CLI = {
    "Linux (32 bits)": "https://github.com/openshift/origin/releases/download/v1.1.1/openshift-origin-client-tools-v1.1.1-e1d9873-linux-32bit.tar.gz",
    "Linux (64 bits)": "https://github.com/openshift/origin/releases/download/v1.1.1/openshift-origin-client-tools-v1.1.1-e1d9873-linux-64bit.tar.gz",
    "Windows":         "https://github.com/openshift/origin/releases/download/v1.1.1/openshift-origin-client-tools-v1.1.1-e1d9873-windows.zip",
    "Mac OS X":        "https://github.com/openshift/origin/releases/download/v1.1.1/openshift-origin-client-tools-v1.1.1-e1d9873-mac.zip"
}
```

spadgett commented:
Fixes #3493

jwforres commented:
and https://github.com/openshift/origin/issues/4344

jwforres commented:
@fabianofranz i think your proposal makes sense, it should be OPENSHIFT_CONSTANTS instead, this aligns with how we namespaced our existing OPENSHIFT_CONFIG

I'm not sure about putting the resources and annotation maps in there, what use case to do you see for someone wanting to override those maps.

fabianofranz commented:
@jwforres ok, changed to `OPENSHIFT_CONSTANTS`. 

About resources and annotations, the idea is to have a central place for these kinds of constants, apart they being subject to overriding or not. Seems more organized than having them right in the body of different filters. Makes sense?

spadgett commented:
@fabianofranz I think it makes sense to keep the constant with the filter if the it's only used by that filter.

Agree with @jwforres: if there are other constants in the file, it looks like they're all customizable.

If a constant is used many places in the web console and shouldn't be configurable, an Angular constant seems appropriate.

fabianofranz commented:
@spadgett ok, got it. I'll remove resources and annotations from the `OPENSHIFT_CONSTANTS` file. I was going to argue you could still use some of them outside the filters (for example, the annotations names in labels) but you could still do that with a filter (`annotationName` filter in that case). Can't think of other cases.

fabianofranz commented:
@jwforres @spadgett fixed, thanks! What else?

jwforres commented on assets/app/scripts/controllers/about.js:
```diff
@@ -0,0 +1,25 @@
+ */
+angular.module('openshiftConsole')
+  .controller('AboutController', function ($scope, DataService, AuthService, Constants) {
+  	$scope.version = {
```
clean up the whitespace for this block, a lot of this is off by one space

fabianofranz replied:
Weird, fixed.

jwforres commented on assets/app/views/about.html:
```diff
@@ -0,0 +1,65 @@
+oc login {{loginBaseURL}} --token={{sessionToken}}
+</pre>
+                <pre class="code prettyprint ng-binding" ng-if="!sessionToken">
+oc login {{loginBaseURL}} --token={{sessionToken}}
```
is this right, you dont have sessionToken in this case

liggitt replied:
I don't really want to expose the token without a click. Can we do the same thing we do for the webhook secret urls with a click to reveal and a copy to clipboard action?

liggitt replied:
yeah, in this case, you'd just have `oc login {{loginBaseURL}}`

fabianofranz replied:
Meh, good catch, fixed the case where we don't have `sessionToken`. Still have to address click-to-reveal.

fabianofranz replied:
Hrm, this is a `<pre`... should we click-to-reveal the entire box? I was thinking about something like below (where "show" is the link), but we will need to use a `div` instead of `pre` and apply the styles.

```
oc login https://localhost:8443 --token=<show>
```

fabianofranz replied:
Ok, fixed to have a click-to-reveal the token.

jwforres commented:
delete the java/bindata.go i keep getting it showing up as a diff for me as well, but it shouldn't be

jwforres commented:
minor stuff, LGTM but if @liggitt has time to double check the go changes that would be good

liggitt commented on pkg/assets/handlers.go:
```diff
@@ -141,6 +144,18 @@ func HTML5ModeHandler(contextRoot string, subcontextMap map[string]string, h htt
+
+type WebConsoleVersion struct {
+	KubernetesVersionInfo kversion.Info
+	OpenShiftVersionInfo  oversion.Info
```
don't pass the whole version struct to the template, just put the version string in the struct given to the template. that removes the need to make this package depend on the version packages, and gives us compiler checks on the use of the GitVersion field

fabianofranz replied:
Nice, fixed.

liggitt commented on pkg/cmd/server/origin/asset.go:
```diff
@@ -209,8 +210,14 @@ func (c *AssetConfig) addHandlers(mux *http.ServeMux) error {
+	oVersionInfo := oversion.Get()
+	versionInfo := assets.WebConsoleVersion{
+		KubernetesVersionInfo: kVersionInfo,
+		OpenShiftVersionInfo:  oVersionInfo,
```
do `OpenShiftVersion: oVersionInfo.GitVersion` here, etc

liggitt commented on pkg/assets/handlers.go:
```diff
@@ -141,6 +144,18 @@ func HTML5ModeHandler(contextRoot string, subcontextMap map[string]string, h htt
+  openshift: "{{ .OpenShiftVersionInfo.GitVersion | js}}",
+  kubernetes: "{{ .KubernetesVersionInfo.GitVersion | js}}"
+};
+`))
```
I would replicate this structure in the static `config.js` and give it a string like "dev-source" or something... just want to keep the `serve-local-assets.sh` env as close as possible to the prod one

fabianofranz replied:
Yeah, I had it originally and removed for cleanliness, but having similar environments is a good argument. Fixed.

fabianofranz commented:
Waiting to rebase when the console ui restructure lands.

fabianofranz commented:
Rebased, [merge].

spadgett commented:
@fabianofranz looks like you need to clean / install assets

jwforres commented:
@fabianofranz java/bindata.go is wrong looks like it needs to be updated, it just got updated yesterday, in general you shouldn't ever be checking in java/bindata.go

fabianofranz commented:
@jwforres meh, I removed and didn't notice it was there again. :/

jwforres commented:
[merge]
